### PR TITLE
added npm tests

### LINKED ISSUE #37: add tests to verify that npm installing local/globally works as expected

### PR DESCRIPTION
**Pull request by rlgomes**
fixes #37

use a docker container to verify the local build can be installed both
globally and locally through NPM and that you can use the available
command line tools to do various tasks.

### CODE REVIEW

**Comment by rlgomes** [comment]
@demmer @mstemm @davidvgalbraith picked a few victims to review... basically this is a few tests that use a docker container to verify that npm installing globally and locally works as expected for juttle-engine and leaves the right bits an pieces in the desired locations.

**Comment by mstemm** [comment]
Can we not run these tests under the general gulp test task? Although useful for travis I think it would be a bit slow to have them run every time a dev wants to run their own unit tests.

File: gulpfile.js [hunk 1]
```diff
@@ -39,7 +39,7 @@ function gulp_test() {
     return gulp.src(tests)
     .pipe(mocha({
         log: true,
-        timeout: 60000,
+        timeout: 1200000,
```
Comment (mstemm) [inline gulpfile.js]: it can really take up to 12 minutes for this to complete?

File: test/npm/npm.spec.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            } else {
+                reject(Error('command: "' + command + ' ' + args.join(' ') +
+                             '", failed with ' + code));
+            }
+        });
+    });
+}
+
+function dockerBash(command) {
+    return spawnify('docker',
```
Comment (mstemm) [inline test/npm/npm.spec.js]: At least for the ubuntu vms I've used, docker commands all require sudo. How will you handle cases where people run this test on linux?

File: gulpfile.js (hunk 1 above)
Comment (rlgomes) [inline gulpfile.js]: 120s is 2 minutes and thats just per test giving it a bit more time to download the docker container + start it up and run the npm install. If you do a fresh npm install from scratch you'll see that alone takes 60s on average.

Comment (mstemm) [inline gulpfile.js]: Oh you're right. I think you're off by 10 then though, you have 1200000 ms / 1200 sec == 20 mins.

File: test/npm/npm.spec.js (hunk 2 above)
Comment (rlgomes) [inline test/npm/npm.spec.js]: docker commands do not require sudo if you follow the last few instructions that docker tells you after each install: 

```If you would like to use Docker as a non-root user, you should now consider
	adding your user to the "docker" group with something like:

	  sudo usermod -aG docker $your_user

	Remember that you will have to log out and back in for this to take effect!
```
Which is the right way to run docker.

**Comment by mstemm** [comment]
Oh also, I would imagine many devs won't have docker-engine installed, in which case these tests won't pass. Another good reason to have them under a separate gulp task.

**Comment by mstemm** [comment]
The tests don't really verify that the command for the docker containers is defined properly, as you override what is run via the docker run command and do all the actual running in docker exec commands. I think to get a good end-to-end test you'd want to verify that just doing a vanilla docker run will work.

File: gulpfile.js (hunk 1 above)
Comment (rlgomes) [inline gulpfile.js]: I'l put a comment on the line indicating the exact time since when units are missing our brains do funny things with too many zeros.

**Comment by rlgomes** [comment]
@mstemm I can split out the "docker dependent" test to something that you'd have to do `gulp test --system` which would kick off system tests including the browser tests as well and by default `gulp test` only runs the unit tests.

**Comment by rlgomes** [comment]
@mstemm I don't follow your second comment...

**Comment by rlgomes** [comment]
@mstemm I don't get this statement:

> The tests don't really verify that the command for the docker containers is defined properly, as you override what is run via the docker run command and do all the actual running in docker exec commands. I think to get a good end-to-end test you'd want to verify that just doing a vanilla docker run will work.

The tests run the global binary and the local binary for each install verifying it can reach those. These isn't testing the existing docker containers we're shipping but instead testing the npm installation of our juttle-engine in the context of a global and local install.

**Comment by mstemm** [comment]
@rlgomes and I discussed in person, I understand the strategy for docker exec vs run now.

**Comment by rlgomes** [comment]
bump!

**Comment by mstemm** [comment]
What did you think about not running this under the regular gulp test task?

**Comment by rlgomes** [comment]
@demmer had said we should run everything under the main gulp task. I'm not against putting this behind flag so `npm test --all/sys` would run all the tests while `npm test` only runs the unit tests that don't need selenium + docker + other adapters + etc.

**Comment by mstemm** [comment]
Ok works for me. I'd definitely like to skip these tests by default.

**Comment by rlgomes** [comment]
bump shall I merge this ? I've added the flag `--sys` and documented its usage in the `test/README.md` then I can basically push all system tests that rely on docker containers for testing to hide behind the same flag and only run in travis by default.

**Comment by mstemm** [comment]
The changes look good to me!

**Pull request merged**